### ISSUE TITLE
[Bug]: Colorbar with log scales wrong format

### ISSUE THREAD

**Issue by RemDelaporteMathurin** [matplotlib/matplotlib#23389]
### Bug summary

The colour bar ticks are weirdly formatted when using a logarithmic scale when using version 3.5.0 or later.

### Code for reproduction

```python
Taken from https://matplotlib.org/stable/gallery/images_contours_and_fields/contourf_log.html


import matplotlib.pyplot as plt
import numpy as np
from numpy import ma
from matplotlib import ticker, cm

N = 100
x = np.linspace(-3.0, 3.0, N)
y = np.linspace(-2.0, 2.0, N)

X, Y = np.meshgrid(x, y)

# A low hump with a spike coming out.
# Needs to have z/colour axis on a log scale so we see both hump and spike.
# linear scale only shows the spike.
Z1 = np.exp(-X**2 - Y**2)
Z2 = np.exp(-(X * 10)**2 - (Y * 10)**2)
z = Z1 + 50 * Z2

# Put in some negative values (lower left corner) to cause trouble with logs:
z[:5, :5] = -1

# The following is not strictly essential, but it will eliminate
# a warning.  Comment it out to see the warning.
z = ma.masked_where(z <= 0, z)


# Automatic selection of levels works; setting the
# log locator tells contourf to use a log scale:
fig, ax = plt.subplots()
cs = ax.contourf(X, Y, z, locator=ticker.LogLocator(), cmap=cm.PuBu_r)

cbar = fig.colorbar(cs)

plt.show()
```

### Actual outcome

![image](https://user-images.githubusercontent.com/40028739/177347713-40db322e-65a3-4763-af66-8c8bb3daab0c.png)

### Expected outcome

Running the exact same code on 3.4.2 produces the expected behaviour
![image](https://user-images.githubusercontent.com/40028739/177347830-9550a81c-7340-49f8-903e-6d39637493f4.png)

### Additional information

This worked perfectly on version 3.4.2

### Operating system

Windows

### Matplotlib Version

>3.5.0

### Installation

pip

**Comment by jklymak** [comment]
Agree this is broken inadvertently.  We refactored some things in colorbar, but the infinite API complexity that crept in over the years is causing problems here again.  I'll try to fix, but we almost need a separate colorbar for contour, rather than tests all over the place.

**Comment by jklymak** [comment]
BTW, a workaround here is to simply call

```python
cbar.ax.set_yscale('log')
```
at the end of the script.

**Comment by RemDelaporteMathurin** [comment]
Interesting, I shall try it out, thanks!

**Comment by adrianterronesaragon** [comment]
> ### Bug summary
> The colour bar ticks are weirdly formatted when using a logarithmic scale when using version 3.5.0 or later.
> 
> ### Code for reproduction
> ```python
> Taken from https://matplotlib.org/stable/gallery/images_contours_and_fields/contourf_log.html
> 
> 
> import matplotlib.pyplot as plt
> import numpy as np
> from numpy import ma
> from matplotlib import ticker, cm
> 
> N = 100
> x = np.linspace(-3.0, 3.0, N)
> y = np.linspace(-2.0, 2.0, N)
> 
> X, Y = np.meshgrid(x, y)
> 
> # A low hump with a spike coming out.
> # Needs to have z/colour axis on a log scale so we see both hump and spike.
> # linear scale only shows the spike.
> Z1 = np.exp(-X**2 - Y**2)
> Z2 = np.exp(-(X * 10)**2 - (Y * 10)**2)
> z = Z1 + 50 * Z2
> 
> # Put in some negative values (lower left corner) to cause trouble with logs:
> z[:5, :5] = -1
> 
> # The following is not strictly essential, but it will eliminate
> # a warning.  Comment it out to see the warning.
> z = ma.masked_where(z <= 0, z)
> 
> 
> # Automatic selection of levels works; setting the
> # log locator tells contourf to use a log scale:
> fig, ax = plt.subplots()
> cs = ax.contourf(X, Y, z, locator=ticker.LogLocator(), cmap=cm.PuBu_r)
> 
> cbar = fig.colorbar(cs)
> 
> plt.show()
> ```
> 
> ### Actual outcome
> ![image](https://user-images.githubusercontent.com/40028739/177347713-40db322e-65a3-4763-af66-8c8bb3daab0c.png)
> 
> ### Expected outcome
> Running the exact same code on 3.4.2 produces the expected behaviour ![image](https://user-images.githubusercontent.com/40028739/177347830-9550a81c-7340-49f8-903e-6d39637493f4.png)
> 
> ### Additional information
> This worked perfectly on version 3.4.2
> 
> ### Operating system
> Windows
> 
> ### Matplotlib Version
> > 3.5.0
> 
> ### Matplotlib Backend
> _No response_
> 
> ### Python version
> _No response_
> 
> ### Jupyter version
> _No response_
> 
> ### Installation
> pip

This issue is still happening in version 3.9.12

**Comment by tacaswell** [comment]
We don't have a 3.9.12, did you mean 3.9.1 or 3.9.2?

I can not reproduce this with 3.9.2, 3.8.4, or main

![so](https://github.com/user-attachments/assets/6cdac781-38a2-477b-840a-2b468ccce72b)

**Comment by adrianterronesaragon** [comment]
Sorry, that is my Python version, my bad. I'm using Matplotlib version 3.5.1.

**Comment by tacaswell** [comment]
This bug looks like it was fixed in mpl 3.6.3 so you will need to upgrade atleast that far to get the fix.
